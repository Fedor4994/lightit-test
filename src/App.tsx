import { Routes, Route } from "react-router-dom";
import Container from "./components/Container/Container";
import Layout from "./views/Layout/Layout";
import ProductsPage from "./views/ProductsPage/ProductsPage";
import LoginPage from "./views/LoginPage/LoginPage";
import ProductDetailsPage from "./views/ProductDetailsPage/ProductDetailsPage";

function App() {
  return (
    <Container>
      <Routes>
        <Route path="/" element={<Layout />}>
          <Route index element={<ProductsPage />} />
          <Route path="/login" element={<LoginPage />} />
          <Route path="/products/:productId" element={<ProductDetailsPage />} />
        </Route>
      </Routes>
    </Container>
  );
}

export default App;
