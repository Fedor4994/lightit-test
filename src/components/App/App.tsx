import { useEffect } from "react";
import { Routes, Route } from "react-router-dom";
import { useSelector } from "react-redux";
import { Toaster } from "react-hot-toast";

import { useAppDispatch } from "../../redux/store";
import { selectIsFetchingUser } from "../../redux/auth/auth-selectors";
import { getCurrentUser } from "../../redux/auth/auth-operations";

import PublicRoute from "../PublicRoute";
import PrivateRoute from "../PrivateRoute";
import Container from "../Container/Container";
import Layout from "../../views/Layout/Layout";
import ProductsPage from "../../views/ProductsPage/ProductsPage";
import LoginPage from "../../views/LoginPage/LoginPage";
import ProductDetailsPage from "../../views/ProductDetailsPage/ProductDetailsPage";
import ReviewsPage from "../../views/ReviewsPage/ReviewsPage";

function App() {
  const dispatch = useAppDispatch();
  const isFetchingUser = useSelector(selectIsFetchingUser);

  useEffect(() => {
    dispatch(getCurrentUser());
  }, [dispatch]);

  return isFetchingUser ? (
    <h1>LOADING...</h1>
  ) : (
    <Container>
      <Routes>
        <Route path="/" element={<Layout />}>
          <Route index element={<ProductsPage />} />
          <Route path="/products/:productId" element={<ProductDetailsPage />} />

          <Route
            path="/reviews"
            element={
              <PrivateRoute component={<ReviewsPage />} redirectTo="/login" />
            }
          />

          <Route
            path="/login"
            element={<PublicRoute component={<LoginPage />} redirectTo="/" />}
          />
        </Route>
      </Routes>
      <Toaster />
    </Container>
  );
}

export default App;
