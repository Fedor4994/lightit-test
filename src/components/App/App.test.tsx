import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import * as reduxHooks from "react-redux";

import App from "./App";

const mockDispatch = jest.fn();
const mockSelector = jest.fn();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
  useSelector: () => mockSelector,
}));

const mockedUseSelector = jest.spyOn(reduxHooks, "useSelector");

describe("App component", () => {
  it("App while fetching current user should contain loader", () => {
    mockedUseSelector.mockReturnValue(true);

    render(<App />);

    const loaderElement = screen.getByText(/loading/i);
    expect(loaderElement).toBeInTheDocument();
  });

  it("App after fetching current user should match snapshot", () => {
    mockedUseSelector.mockReturnValue(false);

    const view = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );

    expect(view).toMatchSnapshot();
  });
});
