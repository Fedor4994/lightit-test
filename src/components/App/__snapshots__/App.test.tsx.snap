// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App after fetching current user should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    data-theme="light"
  >
    <div>
      <div
        class="container"
      >
        <div
          class="content"
        >
          <header
            class="header"
          >
            <a
              class="logo"
              href="/"
            >
              <svg
                class="logoIcon"
                fill="currentColor"
                height="46"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 616 512"
                width="46"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M602 118.6L537.1 15C531.3 5.7 521 0 510 0H106C95 0 84.7 5.7 78.9 15L14 118.6c-33.5 53.5-3.8 127.9 58.8 136.4 4.5.6 9.1.9 13.7.9 29.6 0 55.8-13 73.8-33.1 18 20.1 44.3 33.1 73.8 33.1 29.6 0 55.8-13 73.8-33.1 18 20.1 44.3 33.1 73.8 33.1 29.6 0 55.8-13 73.8-33.1 18.1 20.1 44.3 33.1 73.8 33.1 4.7 0 9.2-.3 13.7-.9 62.8-8.4 92.6-82.8 59-136.4zM529.5 288c-10 0-19.9-1.5-29.5-3.8V384H116v-99.8c-9.6 2.2-19.5 3.8-29.5 3.8-6 0-12.1-.4-18-1.2-5.6-.8-11.1-2.1-16.4-3.6V480c0 17.7 14.3 32 32 32h448c17.7 0 32-14.3 32-32V283.2c-5.4 1.6-10.8 2.9-16.4 3.6-6.1.8-12.1 1.2-18.2 1.2z"
                />
              </svg>
              <span
                class="logoText"
              >
                STUFFSTORE
              </span>
            </a>
            <div
              class="loginIcon"
            >
              <a
                href="/login"
              >
                <svg
                  class="infoIcon"
                  fill="currentColor"
                  height="18"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M313.6 304c-28.7 0-42.5 16-89.6 16-47.1 0-60.8-16-89.6-16C60.2 304 0 364.2 0 438.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-25.6c0-74.2-60.2-134.4-134.4-134.4zM400 464H48v-25.6c0-47.6 38.8-86.4 86.4-86.4 14.6 0 38.3 16 89.6 16 51.7 0 74.9-16 89.6-16 47.6 0 86.4 38.8 86.4 86.4V464zM224 288c79.5 0 144-64.5 144-144S303.5 0 224 0 80 64.5 80 144s64.5 144 144 144zm0-240c52.9 0 96 43.1 96 96s-43.1 96-96 96-96-43.1-96-96 43.1-96 96-96z"
                  />
                </svg>
              </a>
            </div>
          </header>
          <main>
            <div
              class="productsPage"
            >
              <div>
                <div
                  class="filter"
                >
                  Filter
                </div>
              </div>
              <div>
                <div
                  class="productsList"
                />
                <div
                  class="pagination"
                >
                  <button
                    class="paginationButton"
                    disabled=""
                    style="opacity: 0.5; cursor: default;"
                  >
                    <svg
                      fill="currentColor"
                      height="16"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 320 512"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                      />
                    </svg>
                  </button>
                  <div
                    class="currentPage"
                  >
                    1
                  </div>
                  <button
                    class="paginationButton"
                    disabled=""
                    style="opacity: 0.5; cursor: default;"
                  >
                    <svg
                      fill="currentColor"
                      height="16"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 320 512"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </main>
        </div>
        <footer
          class="footer"
        >
          <ul
            class="footerList"
          >
            <li>
              <a
                class="footerIcon"
                href="https://github.com/Fedor4994"
                rel="noopender noreferrer"
                target="_blank"
              >
                <svg
                  fill="currentColor"
                  height="16"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 640 512"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M278.9 511.5l-61-17.7c-6.4-1.8-10-8.5-8.2-14.9L346.2 8.7c1.8-6.4 8.5-10 14.9-8.2l61 17.7c6.4 1.8 10 8.5 8.2 14.9L293.8 503.3c-1.9 6.4-8.5 10.1-14.9 8.2zm-114-112.2l43.5-46.4c4.6-4.9 4.3-12.7-.8-17.2L117 256l90.6-79.7c5.1-4.5 5.5-12.3.8-17.2l-43.5-46.4c-4.5-4.8-12.1-5.1-17-.5L3.8 247.2c-5.1 4.7-5.1 12.8 0 17.5l144.1 135.1c4.9 4.6 12.5 4.4 17-.5zm327.2.6l144.1-135.1c5.1-4.7 5.1-12.8 0-17.5L492.1 112.1c-4.8-4.5-12.4-4.3-17 .5L431.6 159c-4.6 4.9-4.3 12.7.8 17.2L523 256l-90.6 79.7c-5.1 4.5-5.5 12.3-.8 17.2l43.5 46.4c4.5 4.9 12.1 5.1 17 .6z"
                  />
                </svg>
                <span>
                  GitHub
                </span>
              </a>
            </li>
            <li>
              <button
                class="themeSwitcher"
              >
                <svg
                  fill="currentColor"
                  height="16"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                  />
                </svg>
                <span>
                  DARK
                </span>
              </button>
            </li>
          </ul>
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <header
          class="header"
        >
          <a
            class="logo"
            href="/"
          >
            <svg
              class="logoIcon"
              fill="currentColor"
              height="46"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 616 512"
              width="46"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M602 118.6L537.1 15C531.3 5.7 521 0 510 0H106C95 0 84.7 5.7 78.9 15L14 118.6c-33.5 53.5-3.8 127.9 58.8 136.4 4.5.6 9.1.9 13.7.9 29.6 0 55.8-13 73.8-33.1 18 20.1 44.3 33.1 73.8 33.1 29.6 0 55.8-13 73.8-33.1 18 20.1 44.3 33.1 73.8 33.1 29.6 0 55.8-13 73.8-33.1 18.1 20.1 44.3 33.1 73.8 33.1 4.7 0 9.2-.3 13.7-.9 62.8-8.4 92.6-82.8 59-136.4zM529.5 288c-10 0-19.9-1.5-29.5-3.8V384H116v-99.8c-9.6 2.2-19.5 3.8-29.5 3.8-6 0-12.1-.4-18-1.2-5.6-.8-11.1-2.1-16.4-3.6V480c0 17.7 14.3 32 32 32h448c17.7 0 32-14.3 32-32V283.2c-5.4 1.6-10.8 2.9-16.4 3.6-6.1.8-12.1 1.2-18.2 1.2z"
              />
            </svg>
            <span
              class="logoText"
            >
              STUFFSTORE
            </span>
          </a>
          <div
            class="loginIcon"
          >
            <a
              href="/login"
            >
              <svg
                class="infoIcon"
                fill="currentColor"
                height="18"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M313.6 304c-28.7 0-42.5 16-89.6 16-47.1 0-60.8-16-89.6-16C60.2 304 0 364.2 0 438.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-25.6c0-74.2-60.2-134.4-134.4-134.4zM400 464H48v-25.6c0-47.6 38.8-86.4 86.4-86.4 14.6 0 38.3 16 89.6 16 51.7 0 74.9-16 89.6-16 47.6 0 86.4 38.8 86.4 86.4V464zM224 288c79.5 0 144-64.5 144-144S303.5 0 224 0 80 64.5 80 144s64.5 144 144 144zm0-240c52.9 0 96 43.1 96 96s-43.1 96-96 96-96-43.1-96-96 43.1-96 96-96z"
                />
              </svg>
            </a>
          </div>
        </header>
        <main>
          <div
            class="productsPage"
          >
            <div>
              <div
                class="filter"
              >
                Filter
              </div>
            </div>
            <div>
              <div
                class="productsList"
              />
              <div
                class="pagination"
              >
                <button
                  class="paginationButton"
                  disabled=""
                  style="opacity: 0.5; cursor: default;"
                >
                  <svg
                    fill="currentColor"
                    height="16"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 320 512"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                    />
                  </svg>
                </button>
                <div
                  class="currentPage"
                >
                  1
                </div>
                <button
                  class="paginationButton"
                  disabled=""
                  style="opacity: 0.5; cursor: default;"
                >
                  <svg
                    fill="currentColor"
                    height="16"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 320 512"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </main>
      </div>
      <footer
        class="footer"
      >
        <ul
          class="footerList"
        >
          <li>
            <a
              class="footerIcon"
              href="https://github.com/Fedor4994"
              rel="noopender noreferrer"
              target="_blank"
            >
              <svg
                fill="currentColor"
                height="16"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 640 512"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M278.9 511.5l-61-17.7c-6.4-1.8-10-8.5-8.2-14.9L346.2 8.7c1.8-6.4 8.5-10 14.9-8.2l61 17.7c6.4 1.8 10 8.5 8.2 14.9L293.8 503.3c-1.9 6.4-8.5 10.1-14.9 8.2zm-114-112.2l43.5-46.4c4.6-4.9 4.3-12.7-.8-17.2L117 256l90.6-79.7c5.1-4.5 5.5-12.3.8-17.2l-43.5-46.4c-4.5-4.8-12.1-5.1-17-.5L3.8 247.2c-5.1 4.7-5.1 12.8 0 17.5l144.1 135.1c4.9 4.6 12.5 4.4 17-.5zm327.2.6l144.1-135.1c5.1-4.7 5.1-12.8 0-17.5L492.1 112.1c-4.8-4.5-12.4-4.3-17 .5L431.6 159c-4.6 4.9-4.3 12.7.8 17.2L523 256l-90.6 79.7c-5.1 4.5-5.5 12.3-.8 17.2l43.5 46.4c4.5 4.9 12.1 5.1 17 .6z"
                />
              </svg>
              <span>
                GitHub
              </span>
            </a>
          </li>
          <li>
            <button
              class="themeSwitcher"
            >
              <svg
                fill="currentColor"
                height="16"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                />
              </svg>
              <span>
                DARK
              </span>
            </button>
          </li>
        </ul>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
